### PR TITLE
TIF orders

### PR DESCRIPTION
**Pull request by kelvinlau20100**

### CODE REVIEW

File: src/exchange.ts [hunk 1]
```diff
@@ -420,12 +420,6 @@ export class Exchange {
       })
     );
 
-    await Promise.all(
```
Comment (xjcaa) [inline src/exchange.ts:423]: i have context but for @filipzeta - good to comment why

File: src/exchange.ts [hunk 2]
```diff
@@ -570,9 +564,20 @@ export class Exchange {
...
   public async updateAllOrderbooks(live: boolean = true) {
     // This assumes that every market has 1 asksAddress and 1 bidsAddress
-    let allLiveMarkets = this._markets;
+    let allLiveMarkets = [];
+    this.assets.map(async (asset) => {
+      allLiveMarkets = allLiveMarkets.concat(this.getMarkets(asset));
+    });
+
+    for (let i = 0; i < allLiveMarkets.length; i++) {
+      console.log(
```
Comment (xjcaa) [inline src/exchange.ts:573]: todo: remove this later

File: src/exchange.ts [hunk 3]
```diff
@@ -570,9 +564,20 @@ export class Exchange {
 
   public async updateAllOrderbooks(live: boolean = true) {
     // This assumes that every market has 1 asksAddress and 1 bidsAddress
-    let allLiveMarkets = this._markets;
+    let allLiveMarkets = [];
+    this.assets.map(async (asset) => {
+      allLiveMarkets = allLiveMarkets.concat(this.getMarkets(asset));
```
Comment (xjcaa) [inline src/exchange.ts:569]: doesn't this create a copy or is that a ref @kelvinlau20100

File: src/exchange.ts (hunk 1 above)
Comment (filipzeta) [reply src/exchange.ts:423]: is it because subExchange already holds markets?

Comment (kelvinlau20100) [reply src/exchange.ts:423]: so concat actually creates a shallow copy of the markets, now that we actually need up to date market state in some of our utils and when creating the orderbook off chain we need to actually use the markets held in subExchange and not the copy

File: src/exchange.ts (hunk 3 above)
Comment (kelvinlau20100) [reply src/exchange.ts:569]: fine in this situation but have updated code slightly

File: src/client.ts [hunk 4]
```diff
@@ -367,31 +367,23 @@ export class Client {
     price: number,
     size: number,
     side: types.Side,
-    type: types.OrderType = types.OrderType.LIMIT,
-    clientOrderId = 0,
-    tag: String = constants.DEFAULT_ORDER_TAG,
-    blockhash?: string
+    options: types.OrderOptions = types.defaultOrderOptions()
```
Comment (filipzeta) [inline src/client.ts:370]: just noting that this seems to be a breaking change and so should be highlighted in the SDK changelogs

File: src/utils.ts [hunk 5]
```diff
@@ -1630,3 +1645,50 @@ export function getProductLedger(marginAccount: MarginAccount, index: number) {
   }
   return marginAccount.productLedgers[index];
 }
+
+/*
+ TODO: Maybe this should all be done in BN arithmetic because
```
Comment (filipzeta) [inline src/utils.ts:1650]: how large?

File: src/client.ts [hunk 6]
```diff
@@ -401,19 +393,13 @@ export class Client {
...
   ): Promise<TransactionSignature> {
-    return await this.getSubClient(asset).placePerpOrder(
+    return await this.getSubClient(asset).placePerpOrderV2(
```
Comment (xjcaa) [inline src/client.ts:398]: should we also make it placePerp and placeOrder

File: src/exchange.ts (hunk 3 above)
Comment (xjcaa) [reply src/exchange.ts:566]: wait why is it fine?

Comment (xjcaa) [reply src/exchange.ts:566]: cos from what you told me, the markets in allLiveMarkets are copies, so the update orderbook updates the copy, not the original? but if it works then some part of the story is incorrect right

File: src/subclient.ts [hunk 7]
```diff
@@ -610,15 +613,12 @@ export class SubClient {
    * from the user open orders account before cancelling the second order.
    * (Depending on the order in which the order was cancelled).
    */
-  public async placeOrderV3(
+  public async placeOrderV4(
```
Comment (xjcaa) [inline src/subclient.ts:616]: should we nuke vX and just make it placeOrder?

File: src/subclient.ts [hunk 8]
```diff
@@ -687,14 +698,11 @@ export class SubClient {
    * from the user open orders account before cancelling the second order.
    * (Depending on the order in which the order was cancelled).
    */
-  public async placePerpOrder(
+  public async placePerpOrderV2(
```
Comment (xjcaa) [inline src/subclient.ts:701]: just call it placePerpOrder, but make ix v2?

File: src/utils.ts (hunk 5 above)
Comment (xjcaa) [reply src/utils.ts:1650]: yea i feel like it should?

Comment (kelvinlau20100) [reply src/utils.ts:1650]: i gave it more thought i think the only one you would need to do this for is the seqNum comparison, timestamps seem fine to leave as number? but maybe better for consistency

**Review by filipzeta — APPROVED** [review]

**Pull request merged**